Add bookings

LeonovichMariia opened the pull request.

LeonovichMariia committed:
feat: Создан скрипт schema.sql и внесены изменеия в application.properties. Созданы ItemRepository, UserRepository и класс Comment. В классы-сущности добавлены аннотации JPA.
fix: в классы-сервисы доработаны для работы с репозиториями

LeonovichMariia committed:
feat: Созданы дополнительные DTO-классы: UserDtoShort, ItemDtoShort, BookingShortDto, BookingCreationDto. Созданы классы-исключения: BookingException, InvalidIdException, RequestException.
Реализованы функции бронирования и добавлния комментариев: добавлена логика в классы-контроллеры, классы-сервисы и классы хранилища. Добавлены классы: BookingMapper, BookingState и CommentMapper.
fix: Внесены необходимые исправления в классы-сервисы для Item и User для работы с новыми репозиториями.

KirPyt commented on src/main/java/ru/practicum/shareit/booking/dto/BookingMapper.java:
```diff
@@ -0,0 +1,45 @@
+import ru.practicum.shareit.user.model.User;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class BookingMapper {
```
Если хотите - для утилитарного класса так же часто используют аннотацию [@UtilityClass](https://projectlombok.org/features/experimental/UtilityClass)

KirPyt commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,166 @@
+            log.warn(LogMessages.BOOKING_APPROVED.toString());
+            throw new BookingException(LogMessages.BOOKING_APPROVED.toString());
+        }
+        if (approved) {
```
Надо так же проверить, что статус `WAITING`, т.к отклонённую бронь одобрить нельзя

KirPyt commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,166 @@
+    public BookingDto getBookingById(Long bookingId, Long userId) {
+        Booking booking = bookingRepository.validateBooking(bookingId);
+        User user = userRepository.validateUser(userId);
+        if (!(booking.getBooker().getId().equals(user.getId()) || booking.getItem().getOwner().getId().equals(user.getId()))) {
```
Рекомендую вынести выражения в отдельные переменные, так будет лучше читаться

KirPyt commented on src/main/java/ru/practicum/shareit/handler/ErrorHandler.java:
```diff
@@ -36,15 +34,36 @@ public Map<String, String> handleNotFoundException(final NotFoundException e) {
+    public Map<String, String> handleRequestException(final RequestException e) {
+        log.info(LogMessages.BAD_REQUEST_STATUS.toString());
         return Map.of("error", e.getMessage());
     }
```
Надо так же добавить хэндлер на 500-ую ошибку, ловить будем `Throwable`, а возвращать статус `HttpStatus.INTERNAL_SERVER_ERROR`

KirPyt commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -2,78 +2,144 @@
+                .map(ItemMapper::toItemDto)
+                .peek(itemDto -> {
+                    Long itemId = itemDto.getId();
+                    List<Comment> comments = commentRepository.findAllByItemId(itemId);
```
Здесь выходит, что Вы проходитесь по всему списку `findAllByOwnerId`, который имеет длину `N`, а значит далее в вызовах `findAllByItemId` и `setBookings` делаете `N` запросов к БД, надо свести число запросов к минимуму

KirPyt commented on src/main/java/ru/practicum/shareit/user/dto/UserMapper.java:
```diff
@@ -22,4 +22,11 @@ public static User toUser(UserDto userDto) {
                 .build();
     }
+
+//    public static UserDtoShort toUserDtoShort(User user) {
```
Закомментированный код надо убрать

KirPyt reviewed (commented):
Работа очень хорошая, но есть определённые замечания, в частности по запросам к БД, их надо свести к минимуму. В остальном проект выполнен на очень высоком уровне, круто! Так же добавил некоторые мелкие советы по коду

Если будут вопросы, то пишите в Пачку

LeonovichMariia committed:
feat: в класс ErrorHandler добавлены два хендлера handleThrowableException и handleValidationException.

fix: В классы-мапперы добавлена аннотация @UtilityClass.
В классе BookingServiceImpl (в метод renewalBooking) добавлена проверка на WAITING статус.
В классе ItemServiceImpl внесены корректировки в метод getPersonal.

delete: удален закомментированный код

LeonovichMariia committed:
fix: code style

KirPyt reviewed (commented):
Отлично, рад, что у Вас получилось поправить замечания и оптимизировать код)
Ваш проект зачтён
